Add support for managing comments.

piecyk opened the pull request:
This pull request is using much code from awesome work @areina https://github.com/sigma/magit-gh-pulls/pull/26 It's also a work in progress the end goal is to have full support of comments in magit. Adding comments when we are reviewing a commit i magit-diff ( to whole commit or line comment) seeing all comments for a commit. 

I'm just starting my adventure  with `elisp` so help and comments is very needed :+1: please review this madness! 
This can be run locally with the https://github.com/sigma/gh.el/pull/55 i added there support for  commit comments, that we need to have full information about comments in PR. 

next i will add actions for add/edti/delete a comment regarding from where we have it 
(issue, pull, commit comment)

piecyk commented:
Had some time to play with

![image](https://cloud.githubusercontent.com/assets/82964/11693694/9c070a08-9ea6-11e5-8f77-607ae94dda18.png)

regarding the changes to `gh.el` i removed the PR with support of comments much of changes was in the `next` branch... to which i switch small changes to add some missing endpoints and we are back on track... https://github.com/piecyk/gh.el/commit/9636d19388a1fe6eb0ee24aa14b4c539d0347b29 when i got some working version will make proper pr to next branch.

now started to work with the main viewing of comments buffer we need to have toggle action on comment and hide comments that are not from today or stuff ( also show number with how many comments where added today next to overall number ) 

if some is still interested ping me :) and if it's to much noise i will close this scary thing :dancers:

alexander-yakushev commented:
Wow, this look really neat! I would love to merge this when it's stable and polished.

Now, are you saying that this functionality requries `gh.el` from `next` branch? Then I hope Yann merges it with master soon, so that it becomes available in Melpa.

piecyk commented:
@alexander-yakushev thanks, it will look better :+1: 

Regarding the `gh.el` i submitted a pr to master but then realized that much of this is done i `next` branch so to don't duplicate the work i closed that pr, and switch to next branch... When i got something stable and will be happy with code will submit a pr witch that changes needed.

alexander-yakushev commented:
Great, looking forward to it! I'll try to give the code itself a look sometime on the weekend.

piecyk commented:
had few minutes for this, so basic change is to use magit-modes ect... 
still we need better formatting, including the diff if comment on line in pull or commit// 

![c](https://cloud.githubusercontent.com/assets/82964/11768305/8a17cfb8-a1c0-11e5-979f-29b487fa87d5.gif)

alexander-yakushev commented:
Amazing!

I think the PR comments should be unfolded by default and be on top. Also, it would **super** cool if we cold show the diff chunk for commit/PR comments. Do you think it is possible?

piecyk commented:
It's out of box because github api have the `diff-chunk`, next step would be to make it like in `magit-diff` i added to show the diff ( default hidden ) also added popup menu for action to go to file ( but we need to get the line from the diff ) now it will only go to correct file and commit 

![image](https://cloud.githubusercontent.com/assets/82964/11823661/8052da70-a375-11e5-869d-a8cc00449b8e.png)

sigma commented:
@piecyk that looks pretty good. I'm looking forward to seeing your changes in gh.el next :)

As for comment->diff navigation, I'd personally find it awesome if we could jump to the right chunk in the global diff instead of/in addition to inlining that chunk in the comment section (and being able to come back would be even greater)
Thanks !

norcalli commented:
Since the `gh.el` changes have been merged, can this be updated now to work? Even if it's just a beta version, it would still be useful.

danielmartin commented:
What's the status of this? I'd love to continue hacking on this feature if the author does not have enough free time to complete it.

jb55 commented:
Looks amazing. Very close to the dream of never having to leave magit. Reading PRs, comments, eventually code review!? Let's not let dreams be dreams.
